File: solutions/0787. Cheapest Flights Within K Stops/0787.cpp
struct T {
  int u;
  int d;
  int stops;
  T(int u, int d, int stops) : u(u), d(d), stops(stops) {}
};

class Solution {
 public:
  int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst,
                        int k) {
    vector<vector<pair<int, int>>> graph(n);
    auto compare = [](const T& a, const T& b) { return a.d > b.d; };
    priority_queue<T, vector<T>, decltype(compare)> minHeap(compare);
    vector<vector<int>> dist(n, vector<int>(k + 2, INT_MAX));

    minHeap.emplace(src, 0, k + 1);  // start with node src with d == 0
    dist[src][k + 1] = 0;

    for (const auto& f : flights) {
      const int u = f[0];
      const int v = f[1];
      const int w = f[2];
      graph[u].emplace_back(v, w);
    }

    while (!minHeap.empty()) {
      const auto [u, d, stops] = minHeap.top();
      minHeap.pop();
      if (u == dst)
        return d;
      if (stops > 0)
        for (const auto& [v, w] : graph[u]) {
          const int newDist = d + w;
          if (newDist < dist[v][stops - 1]) {
            dist[v][stops - 1] = newDist;
            minHeap.emplace(v, newDist, stops - 1);
          }
        }
    }

    return -1;
  }
};

File: solutions/0787. Cheapest Flights Within K Stops/0787.java
class T {
  public int u;
  public int d;
  public int stops;
  public T(int u, int d, int stops) {
    this.u = u;
    this.d = d;
    this.stops = stops;
  }
}

class Solution {
  public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
    List<Pair<Integer, Integer>>[] graph = new List[n];
    PriorityQueue<T> minHeap = new PriorityQueue<>((a, b) -> a.d - b.d);
    int[][] dist = new int[n][k + 2];
    Arrays.stream(dist).forEach(A -> Arrays.fill(A, Integer.MAX_VALUE));

    for (int i = 0; i < n; ++i)
      graph[i] = new ArrayList<>();

    for (int[] f : flights) {
      final int u = f[0];
      final int v = f[1];
      final int w = f[2];
      graph[u].add(new Pair<>(v, w));
    }

    minHeap.offer(new T(src, 0, k + 1)); // start with node src with d == 0
    dist[src][k + 1] = 0;

    while (!minHeap.isEmpty()) {
      final int u = minHeap.peek().u;
      final int d = minHeap.peek().d;
      final int stops = minHeap.poll().stops;
      if (u == dst)
        return d;
      if (stops > 0)
        for (var node : graph[u]) {
          final int v = node.getKey();
          final int w = node.getValue();
          final int newDist = d + w;
          if (newDist < dist[v][stops - 1]) {
            dist[v][stops - 1] = newDist;
            minHeap.offer(new T(v, d + w, stops - 1));
          }
        }
    }

    return -1;
  }
}
